doc/kwim.1
.\" Generated by scdoc 1.11.4
.\" Complete documentation for this program is not available as a GNU info page
.ie \n(.g .ds Aq \(aq
.el       .ds Aq '
.nh
.ad l
.\" Begin generated content:
.TH "KWIM" "1" "2026-04-19" "kwim" "kwm input manager"
.PP
.SH NAME
.PP
kwim - kwm input manager for River Wayland compositor
.PP
.SH SYNOPSIS
.PP
\fBkwim\fR [OPTION]
.br
\fBkwim\fR \fBSUBCOMMAND\fR \fIDEVICE_TYPE\fR \fIDEVICE_OPTION\fR.\&.\&.\& [ARG].\&.\&.\&
.PP
.SH DESCRIPTION
.PP
\fIkwm\fR is a DWM-like dynamic tiling window manager implementing the
river-window-management-v1 protocol.\&
.PP
\fBkwim\fR is an input manager separated from \fIkwm\fR, implementing the
river-input-management-v1 protocol and/or related protocols in order to
configure input devices independent of window manager.\&
.PP
\fIkwm\fR will automatically run \fBkwim\fR at startup.\& You can also run \fBkwim\fR to list
input devices or apply a single rule on demand.\&
.PP
Without any subcommands and options, \fBkwim\fR reads input rules from the same
configuration file used by \fIkwm\fR.\&
.PP
.SH OPTIONS
.PP
\fB-h,--help\fR
.RS 4
Print a help message and exit.\&
.PP
.RE
\fB-v,--version\fR
.RS 4
Print the version and exit.\&
.PP
.RE
\fB-c,--config\fR
.RS 4
Specify the configuration file path.\&
.PP
.RE
.SH SUBCOMMANDS
.PP
\fBlist\fR
.RS 4
Lists all connected devices of the specified type.\&
.PP
.RE
\fBapply\fR
.RS 4
Applies configuration rules to a specific device type.\&
.PP
.RE
.SH DEVICE\_TYPE
.PP
\fBinput-device\fR
.PP
\fBlibinput-device\fR
.PP
\fBxbk-keyboard\fR
.PP
.SH DEVICE\_OPTIONS
.PP
.SS Common Subcommand Options
.PP
\fB-n,--name\fR \fIDEVICE_NAME\fR
.RS 4
Input device name
.PP
.RE
\fB--regex\fR
.RS 4
Enable regex
.PP
.RE
\fB--match-null\fR
.RS 4
If to match null name
.PP
.RE
.SS input-device options
.PP
\fB--repeat-info\fR \fI<REPEAT_INFO>\fR
.RS 4
Keyboard repeat info (e.\&g.\& 50,300)
.PP
.RE
\fB--scroll-factor\fR \fI<SCROLL_FACTOR>\fR
.RS 4
Pointer scroll factor
.PP
.RE
.SS libinput-device options
.PP
\fB--send-event-modes\fR \fISEND_EVENT_MODE_STATE\fR
.RS 4
Set send events mode (enabled, disabled, disabled_on_external_mouse)
.PP
.RE
\fB--tap\fR \fITAP_STATE\fR
.RS 4
Set tap to click state (enabled, disabled)
.PP
.RE
\fB--drag\fR \fIDRAG_STATE\fR
.RS 4
Set tap-and-drag state (enabled, disabled)
.PP
.RE
\fB--drag-lock\fR \fIDRAG_LOCK_STATE\fR
.RS 4
Set tap-and-drag lock state (enabled, disabled)
.PP
.RE
\fB--tap-button-map\fR \fITAP_BUTTON_MAP\fR
.RS 4
Set tap button map (left_right_middle, left_middle_right)
.PP
.RE
\fB--three-finger-drag\fR \fITHREE_FINGER_DRAG_STATE\fR
.RS 4
Set three finger drag state (enabled, disabled)
.PP
.RE
\fB--calibration-matrix\fR \fICALIBRATION_MATRIX\fR
.RS 4
Set calibration matrix (format: "a,b,c,d,e,f")
.PP
.RE
\fB--accel-profile\fR \fIACCEL_PROFILE\fR
.RS 4
Set acceleration profile (flat, adaptive, custom)
.PP
.RE
\fB--accel-speed\fR \fIACCEL_SPEED\fR
.RS 4
Set acceleration speed (-1.\&0 to 1.\&0)
.PP
.RE
\fB--natural-scroll\fR \fINATURAL_SCROLL_STATE\fR
.RS 4
Set natural scrolling state (enabled, disabled)
.PP
.RE
\fB--left-handed\fR \fILEFT_HANDED_STATE\fR
.RS 4
Set left-handed mode (enabled, disabled)
.PP
.RE
\fB--click-method\fR \fICLICK_METHOD\fR
.RS 4
Set click method (none, button_areas, clickfinger)
.PP
.RE
\fB--clickfinger-button-map\fR \fICLICKFINGER_BUTTON_MAP\fR
.RS 4
Set clickfinger button map (left_right_middle, left_middle_right)
.PP
.RE
\fB--middle-button-emulation\fR \fIMIDDLE_BUTTON_EMULATION_STATE\fR
.RS 4
Set middle button emulation (enabled, disabled)
.PP
.RE
\fB--scroll-method\fR \fISCROLL_METHOD\fR
.RS 4
Set scroll method (none, two_finger, edge, on_button_down,
on_button_down_lock)
.PP
.RE
\fB--scroll-button\fR \fISCROLL_BUTTON\fR
.RS 4
Set scroll button (e.\&g.\&, left, right, middle, side, extra, forward, back,
task)
.PP
.RE
\fB--scroll-button-lock\fR \fISCROLL_BUTTON_LOCK_STATE\fR
.RS 4
Set scroll button lock state (enabled, disabled)
.PP
.RE
\fB--disable-while-typing\fR \fIDISABLE_WHILE_TYPING_STATE\fR
.RS 4
Set disable-while-typing state (enabled, disabled)
.PP
.RE
\fB--disable-while-trackpointing\fR \fIDISABLE_WHILE_TRACKPOINTING_STATE\fR
.RS 4
Set disable-while-trackpointing state (enabled, disabled)
.PP
.RE
\fB--rotation-angle\fR \fIROTATION_ANGLE\fR
.RS 4
Set rotation angle in degrees (0-360)
.PP
.RE
.SS xkb-keyboard options
.PP
\fB--numlock\fR \fINUMLOCK_STATE\fR
.RS 4
Set numlock state (enabled, disabled)
.PP
.RE
\fB--capslock\fR \fICAPSLOCK_STATE\fR
.RS 4
Set capslock state (enabled, disabled)
.PP
.RE
\fB--layout\fR \fIKEYBOARD_LAYOUT\fR
.RS 4
Keyboard layout name, index or name
.PP
.RE
\fB--keymap-file\fR \fISTRING\fR
.RS 4
Keymap file (e.\&g.\& <file-path>@<format>)
.PP
.RE
\fB--keymap-options\fR \fISTRING\fR
.RS 4
Keymap options (e.\&g.\& rules=.\&.\&.\&,model=.\&.\&.\&,layout=.\&.\&.\&,variant=.\&.\&.\&,options=.\&.\&.\&)
.PP
.RE
.SH CONFIGURATION
.PP
By default, kwim reads input rules from the same configuration file used by
\fIkwm\fR, in the following paths:
.PP
.RS 4
$XDG_CONFIG_HOME/kwm/config.\&zon
.PP
$HOME/.\&config/kwm/config.\&zon
.PP
.RE
When user configuration doesn'\&t exists, kwim will try to use the compile time
configuration.\&
.PP
See \fIkwim(5)\fR for all possible settings.\&
.PP
.SH SEE ALSO
.PP
\fIkwim\fR(5), \fIkwm\fR(1), \fIriver\fR(1)
.PP
.SH BUGS
.PP
https://github.\&com/kewuaa/kwim/issues


completions/kwim.bash
_kwim() {
    local cur prev words cword
    _get_comp_words_by_ref -n : cur prev words cword

    local subcommands="list apply"
    local devices="input-device libinput-device xkb-keyboard"
    local options="-h --help -v --version -c --config"

    local common_options="-h --help --name --regex --match-null"
    local input_device_options="--repeat-info --scroll-factor"
    local libinput_options="--send-event-modes --tap --drag --drag-lock --tap-button-map --three-finger-drag --calibration-matrix --accel-profile --accel-speed --natural-scroll --left-handed --click-method --clickfinger-button-map --middle-button-emulation --scroll-method --scroll-button --scroll-button-lock --disable-while-typing --disable-while-trackpointing --rotation-angle"
    local xkb_options="--numlock --capslock --layout --keymap-file --keymap-rules --keymap-model --keymap-layout --keymap-variant --keymap-options"

    local bool_states="enabled disabled"
    local send_event_modes="enabled disabled disabled_on_external_mouse"
    local tap_button_maps="left_right_middle left_middle_right"
    local accel_profiles="flat adaptive custom"
    local click_methods="none button_areas clickfinger"
    local scroll_methods="none two_finger edge on_button_down on_button_down_lock"
    local buttons="left right middle side extra forward back task"
    local numlock_states="enabled disabled"
    local capslock_states="enabled disabled"

    if [[ "$prev" == "-c" || "$prev" == "--config" ]]; then
        _filedir
        return 0
    fi

    if [[ "$prev" == "--name" ]]; then
        return 0
    fi

    case "$prev" in
        --send-event-modes)
            COMPREPLY=( $(compgen -W "$send_event_modes" -- "$cur") )
            return 0
            ;;
        --tap|--drag|--drag-lock|--three-finger-drag|--natural-scroll|--left-handed|--middle-button-emulation|--scroll-button-lock|--disable-while-typing|--disable-while-trackpointing)
            COMPREPLY=( $(compgen -W "$bool_states" -- "$cur") )
            return 0
            ;;
        --tap-button-map|--clickfinger-button-map)
            COMPREPLY=( $(compgen -W "$tap_button_maps" -- "$cur") )
            return 0
            ;;
        --accel-profile)
            COMPREPLY=( $(compgen -W "$accel_profiles" -- "$cur") )
            return 0
            ;;
        --click-method)
            COMPREPLY=( $(compgen -W "$click_methods" -- "$cur") )
            return 0
            ;;
        --scroll-method)
            COMPREPLY=( $(compgen -W "$scroll_methods" -- "$cur") )
            return 0
            ;;
        --scroll-button)
            COMPREPLY=( $(compgen -W "$buttons" -- "$cur") )
            return 0
            ;;
        --numlock)
            COMPREPLY=( $(compgen -W "$numlock_states" -- "$cur") )
            return 0
            ;;
        --capslock)
            COMPREPLY=( $(compgen -W "$capslock_states" -- "$cur") )
            return 0
            ;;
        --calibration-matrix|--repeat-info|--keymap-file|--keymap-rules|--keymap-model|--keymap-layout|--keymap-variant|--keymap-options|--accel-speed|--scroll-factor|--rotation-angle|--layout)
            return 0
            ;;
    esac

    if [[ "$cur" == -* ]]; then
        if [[ ${#words[@]} -eq 2 ]]; then
            COMPREPLY=( $(compgen -W "$options" -- "$cur") )
            return 0
        fi

        if [[ ${#words[@]} -ge 3 ]]; then
            case "${words[1]}" in
                apply)
                    if [[ ${#words[@]} -eq 3 ]]; then
                        COMPREPLY=( $(compgen -W "$devices" -- "$cur") )
                    else
                        case "${words[2]}" in
                            input-device)
                                COMPREPLY=( $(compgen -W "$common_options $input_device_options" -- "$cur") )
                                ;;
                            libinput-device)
                                COMPREPLY=( $(compgen -W "$common_options $libinput_options" -- "$cur") )
                                ;;
                            xkb-keyboard)
                                COMPREPLY=( $(compgen -W "$common_options $xkb_options" -- "$cur") )
                                ;;
                            *)
                                COMPREPLY=( $(compgen -W "$devices" -- "$cur") )
                                ;;
                        esac
                    fi
                    return 0
                    ;;
                list)
                    if [[ ${#words[@]} -eq 3 ]]; then
                        COMPREPLY=( $(compgen -W "$devices" -- "$cur") )
                    else
                        COMPREPLY=( $(compgen -W "$common_options" -- "$cur") )
                    fi
                    return 0
                    ;;
            esac
        fi

        COMPREPLY=( $(compgen -W "$options" -- "$cur") )
        return 0
    fi

    if [[ ${#words[@]} -eq 2 ]]; then
        COMPREPLY=( $(compgen -W "$subcommands" -- "$cur") )
    elif [[ ${#words[@]} -eq 3 ]]; then
        case "${words[1]}" in
            apply|list)
                COMPREPLY=( $(compgen -W "$devices" -- "$cur") )
                ;;
        esac
    elif [[ ${#words[@]} -ge 4 ]]; then
        case "${words[1]}" in
            apply)
                case "${words[2]}" in
                    input-device)
                        COMPREPLY=( $(compgen -W "$common_options $input_device_options" -- "$cur") )
                        ;;
                    libinput-device)
                        COMPREPLY=( $(compgen -W "$common_options $libinput_options" -- "$cur") )
                        ;;
                    xkb-keyboard)
                        COMPREPLY=( $(compgen -W "$common_options $xkb_options" -- "$cur") )
                        ;;
                esac
                ;;
            list)
                case "${words[2]}" in
                    input-device|libinput-device|xkb-keyboard)
                        COMPREPLY=( $(compgen -W "$common_options" -- "$cur") )
                        ;;
                esac
                ;;
        esac
    else
        COMPREPLY=()
    fi
}

complete -F _kwim kwim
